frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/GitFetchSupportImpl.java

```diff
@@ -0,0 +1,462 @@
+        }
+      }
+    }
+    return fetch(remotesToFetch);
+  }
+
+  public FetchResultImpl fetch(GitRepository repository, GitRemote remote, String refspec) {
+    return fetch(List.of(new RemoteRefCoordinates(repository, remote, refspec)));
+  }
+
+  @Override
+  public FetchResultImpl fetch(GitRepository repository, GitRemote remote) {
+    return fetch(List.of(new RemoteRefCoordinates(repository, remote, /* refspec */ null)));
+  }
+
+  public FetchResultImpl fetch(List<RemoteRefCoordinates> arguments) {
+    try {
+      fetchRequestCounter.incrementAndGet();
+      return withIndicator(() -> {
+        var activity = IdeActivity.started(project, "vcs", "fetch");
+
+        var tasks = fetchInParallel(arguments);
+        var results = waitForFetchTasks(tasks);
+
+        var mergedResults = new HashMap<GitRepository, RepoResult>();
+        for (var result : results) {
+          var res = mergedResults.get(result.repository);
+          mergedResults.put(result.repository, mergeRepoResults(res, result));
+        }
+        activity.finished();
+
+        return new FetchResultImpl(project, VcsNotifier.getInstance(project), mergedResults);
+      });
+    } finally {
+      fetchRequestCounter.decrementAndGet();
+    }
+  }
+
+  private List<FetchTask> fetchInParallel(List<RemoteRefCoordinates> remotes) {
+    var tasks = new ArrayList<FetchTask>();
+    var maxThreads = getMaxThreads(remotes.stream().map(r -> r.repository).collect(Collectors.toList()), remotes.size());
+    LOG.debug("Fetching ${remotes} using ${maxThreads} threads");
+    var executor = AppExecutorUtil.createBoundedApplicationPoolExecutor("GitFetch pool", maxThreads);
+    var commonIndicator = Option.of(progressManager.getProgressIndicator()).getOrElse(new EmptyProgressIndicator());
+    var authenticationGate = new GitRestrictingAuthenticationGate();
+    for (var cords : remotes) {
+      LOG.debug("Fetching ${cords.remote} in ${cords.repository}");
+      Future<SingleRemoteResult> future = executor.submit(() -> {
+        commonIndicator.checkCanceled();
+        AtomicReference<@Nullable SingleRemoteResult> result = new AtomicReference<>(null);
+
+        ProgressManager.getInstance().executeProcessUnderProgress(() -> {
+          commonIndicator.checkCanceled();
+          result.set(doFetch(cords.repository, cords.remote, cords.refspec, authenticationGate));
+        }, commonIndicator);
+
+        SingleRemoteResult singleRemoteResult = result.get();
+        assert singleRemoteResult != null : "Single remote result is null";
+        return singleRemoteResult;
+      });
+      tasks.add(new FetchTask(cords.repository, cords.remote, future));
+    }
+    return tasks;
+  }
+
+  private int getMaxThreads(Collection<GitRepository> repositories, int numberOfRemotes) {
+    var config = Registry.intValue("git.parallel.fetch.threads");
+    var maxThreads = 1;
+    if (config > 0) {
+      maxThreads = config;
+    } else if (config == -1) {
```

Comment:
Are those values available somewhere as named constants? :frowning_face:

Reply:
I would like to leave this description from `Registry` as reference here.

```
Number of threads to call git fetch in parallel.
Set to 1 to execute sequentially.
Set to -1 to use the number of CPU cores.
Set to -2 to start as many threads, as there are repositories in the project.
Set to -3 to start as many threads, as there are repositories, but not more than 2 * number of cores.
By default, SSH supports up to 10 multiplexed connections, thus number of threads will be limited by 10 despite this value.
```

Reply:
Actually we can consider using fixed value in our implementation. 
 - it will simplify the logic 
 - our implementation is a temporary workaround

Reply:
Note that the `-3` is set by **default**.

Reply:
_Slack discussion summary_

Use the fixed value `-3`.